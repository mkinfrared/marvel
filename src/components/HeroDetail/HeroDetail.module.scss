.HeroDetail {
  padding: 20px;

  > h2 {
    font-size: 32px;
    font-weight: 800;
    margin-bottom: 20px;
  }

  .image {
    margin: 0 auto;
    margin-bottom: 20px;
    max-width: 500px;
    width: 100%;

    img {
      height: 100%;
      object-fit: contain;
      width: 100%;
    }
  }

  > p {
    font-size: 24px;
    line-height: 1.4;
    text-align: justify;
  }

  .loading {
    align-items: center;
    display: flex;
    justify-content: center;
  }
}
