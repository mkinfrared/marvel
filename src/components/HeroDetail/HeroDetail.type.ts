import { Dispatch } from "@reduxjs/toolkit";

import { fetchHeroRequest } from "store/reducers/heroes/reducer";
import { Hero } from "store/reducers/heroes/types";

export type HeroDetailProps = Partial<Hero> & {
  id: number;
  dispatch: Dispatch<ReturnType<typeof fetchHeroRequest>>;
};
