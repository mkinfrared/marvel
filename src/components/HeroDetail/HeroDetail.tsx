/* eslint-disable no-shadow */
import React, { useEffect, useLayoutEffect, useRef, useState } from "react";

import { fetchHeroRequest } from "store/reducers/heroes/reducer";
import { ComicBook, Event, Series, Story } from "store/reducers/heroes/types";
import Card from "ui/Card";
import Image from "ui/Image";
import Section from "ui/Section";
import Spinner from "ui/Spinner";

import css from "./HeroDetail.module.scss";
import { HeroDetailProps } from "./HeroDetail.type";

type Entity = ComicBook | Event | Series | Story;

const HeroDetail = ({
  comics,
  description,
  events,
  id,
  name,
  series,
  stories,
  thumbnail,
  dispatch
}: HeroDetailProps) => {
  const [width, setWidth] = useState(100);
  const ref = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (!name) {
      dispatch(fetchHeroRequest(id));
    }
  }, [id]);

  useLayoutEffect(() => {
    if (ref.current) {
      setWidth(ref.current.clientWidth);
    }
  }, [name]);

  const getCard = (entity: Entity) => {
    const { id, title, thumbnail } = entity;
    const imageSrc = `${thumbnail?.path}.${thumbnail?.extension}`;

    return <Card key={id} imageSrc={imageSrc} title={title} />;
  };

  const renderData = () => {
    if (!name) {
      return (
        <div className={css.loading}>
          <Spinner />
        </div>
      );
    }

    const imageSrc = `${thumbnail?.path}.${thumbnail?.extension}`;
    const heroComicBooks = comics?.map(getCard);
    const heroEvents = events?.map(getCard);
    const heroSeries = series?.map(getCard);
    const heroStories = stories?.map(getCard);

    return (
      <>
        <h2>{name}</h2>
        <div className={css.image} ref={ref}>
          <Image src={imageSrc} loadWidth={width} loadHeight={width} />
        </div>
        <p>{description}</p>
        <Section title={`${name} comic books`}>{heroComicBooks}</Section>
        <Section title={`${name} events`}>{heroEvents}</Section>
        <Section title={`${name} series`}>{heroSeries}</Section>
        <Section title={`${name} stories`}>{heroStories}</Section>
      </>
    );
  };

  return <div className={css.HeroDetail}>{renderData()}</div>;
};

export default React.memo(HeroDetail);
