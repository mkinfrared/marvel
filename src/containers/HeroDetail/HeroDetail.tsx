import React from "react";
import { useDispatch, useSelector } from "react-redux";

import HeroDetailFC from "components/HeroDetail";
import { HeroDetailProps } from "components/HeroDetail/HeroDetail.type";
import { selectHero } from "store/reducers/heroes/selectors";

const HeroDetail = ({ id }: Pick<HeroDetailProps, "id">) => {
  const hero = useSelector(selectHero(id));
  const dispatch = useDispatch();

  return <HeroDetailFC {...hero} id={id} dispatch={dispatch} />;
};

export default React.memo(HeroDetail);
