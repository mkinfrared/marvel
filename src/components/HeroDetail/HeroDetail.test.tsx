import { render } from "@testing-library/react";
import React from "react";

import {
  comicsMock,
  eventsMock,
  seriesMock,
  storiesMock
} from "store/reducers/heroes/test/heroesMock";

import HeroDetail from "./HeroDetail";

jest.mock("ui/Image", () => () => "<Image />");

describe("<HeroDetail />", () => {
  const id = 42;
  const name = "marklar";
  const dispatch = jest.fn();
  const thumbnail = { path: "foo/bar", extension: "bmp" };
  const comics = comicsMock as any;
  const events = eventsMock;
  const series = seriesMock;
  const stories = storiesMock;
  const props = { id, dispatch };
  const rest = { comics, events, series, stories, thumbnail, name };
  const Component = <HeroDetail {...props} />;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should call dispatch from props when name is not defined", () => {
    render(Component);

    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  it("should not call dispatch from props when name is defined", () => {
    const Comp = <HeroDetail {...props} {...rest} />;
    const { container } = render(Comp);
    const header = container.querySelector("h2");

    expect(dispatch).not.toHaveBeenCalled();

    expect(header).toBeDefined();
  });
});
