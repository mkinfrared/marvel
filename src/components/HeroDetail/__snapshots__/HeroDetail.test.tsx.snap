// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroDetail /> should match the snapshot 1`] = `
<div>
  <div
    class="HeroDetail"
  >
    <div
      class="loading"
    >
      <div
        class="Spinner"
        style="width: 25px; height: 25px;"
      />
    </div>
  </div>
</div>
`;
